### PR TITLE
Feature/node checker remove failed nodes

### PR DESCRIPTION
**Pull request by bdharrington7**
Hi @kramer I fixed up @bmlining's PR, hopefully this works for everyone

### CODE REVIEW

**Comment by bmlining** [comment]
Looks good, even better approach than I was thinking of.

**Comment by bdharrington7** [comment]
Merging this would close #329 and #345

**Comment by kramer** [comment]
NodeChecker is not the only one calling execute therefore losing the cause-exception does not seem like a good idea to me, I would prefer something like below:

```java 
    public CouldNotConnectException(String host, Throwable cause) {
        super("Could not connect to " + host, cause);
        this.host = host;
    }
```

So people still have the details when debugging/researching an odd connection exception.

**Comment by kramer** [comment]
To keep the behaviour of `HttpHostConnectException` being "converted" to `CouldNotConnectException` same also for async execution maybe `DefaultCallback.failed` should also be modified accordingly.

File: jest-common/src/main/java/io/searchbox/client/config/discovery/NodeChecker.java [hunk 1]
```diff
@@ -90,13 +101,23 @@ protected void runOneIteration() throws Exception {
...
+            discoveredServerList = httpHosts;
+            client.setServers(discoveredServerList);
         } else {
             log.warn("NodesInfo request resulted in error: {}", result.getErrorMessage());
             client.setServers(bootstrapServerList);
         }
     }
 
+    private void removeNodeAndUpdateServers(final String hostToRemove) {
+        discoveredServerList.remove(hostToRemove);
```
Comment (kramer) [inline jest-common/src/main/java/io/searchbox/client/config/discovery/NodeChecker.java]: Should info/warn log the taken action.

File: jest-common/src/main/java/io/searchbox/client/config/discovery/NodeChecker.java [hunk 2]
```diff
@@ -90,13 +101,23 @@ protected void runOneIteration() throws Exception {
...
-            client.setServers(httpHosts);
+            discoveredServerList = httpHosts;
+            client.setServers(discoveredServerList);
         } else {
             log.warn("NodesInfo request resulted in error: {}", result.getErrorMessage());
             client.setServers(bootstrapServerList);
         }
     }
 
+    private void removeNodeAndUpdateServers(final String hostToRemove) {
```
Comment (kramer) [inline jest-common/src/main/java/io/searchbox/client/config/discovery/NodeChecker.java]: `protected` maybe for people who want to customize the behaviour?

**Comment by kramer** [comment]
Could you also please adapt `JestDroidClient` accordingly?

**Comment by kramer** [comment]
Thank you for your contribution @bdharrington7 and thank you for the initial effort @bmlining !

**Pull request merged**